core/transformer/src/IModelTransformer.ts

```diff
@@ -102,6 +104,100 @@ export interface IModelTransformOptions {
+ * @internal
+ */
+class PartiallyCommittedElement {
+  public constructor(
+    /**
+     * A set of "model|element ++ ID64" pairs, e.g. `model0x11` or `element0x12`.
+     * It is possible for the submodel of an element to be separately resolved from the actual element,
+     * so its resolution must be tracked separately
+     */
+    private _missingPredecessors: Set<string>,
+    private _onComplete: () => void
+  ) {}
+  public resolvePredecessor(id: Id64String, isModelRef: boolean) {
+    const key = PartiallyCommittedElement.makePredecessorKey(id, isModelRef);
+    this._missingPredecessors.delete(key);
+    if (this._missingPredecessors.size === 0) this._onComplete();
+  }
+  public static makePredecessorKey(id: Id64String, isModelRef: boolean) {
+    return `${isModelRef ? "model" : "element"}${id}`;
+  }
+  public forceComplete() {
+    this._onComplete();
+  }
+}
+
+/**
+ * A helper for checking the in-transformation processing state of an element,
+ * whether it has been transformed, and if its submodel has been
+ * @internal
+ */
+class ElementProcessState {
+  public constructor(
+    public elementId: string,
+    /** whether or not this element needs to be processed */
+    public needsElemImport: boolean,
+    /** whether or not this element's submodel needs to be processed */
+    public needsModelImport: boolean,
+  ) {}
+  public static fromElementAndTransformer(elementId: Id64String, transformer: IModelTransformer): ElementProcessState {
+    // we don't need to load all of the props of the model to check if the model exists
+    const dbHasModel = (db: IModelDb, id: Id64String) => db.withPreparedStatement("SELECT 1 FROM bis.Model WHERE ECInstanceId=? LIMIT 1", (stmt) => {
+      stmt.bindId(1, id);
+      const stepResult = stmt.step();
+      if (stepResult === DbResult.BE_SQLITE_DONE) return false;
+      if (stepResult === DbResult.BE_SQLITE_ROW) return true;
+      else throw new IModelError(stepResult, "expected 1 or no rows");
+    });
+    const isSubModeled = dbHasModel(transformer.sourceDb, elementId);
+    const idOfElemInTarget = transformer.context.findTargetElementId(elementId);
+    const isElemInTarget = Id64.invalid !== idOfElemInTarget;
+    const needsModelImport = isSubModeled && (!isElemInTarget || !dbHasModel(transformer.targetDb, idOfElemInTarget));
+    return new ElementProcessState(elementId, !isElemInTarget, needsModelImport);
+  }
+  public get needsImport() { return this.needsElemImport || this.needsModelImport; }
+}
+
+/**
+ * Apply a function to each Id64 in a supported container type of Id64s.
+ * Currently only supports raw Id64String or RelatedElement-like objects containing an `id` property that is a Id64String,
+ * which matches the possible containers of references in [Element.requiredReferenceKeys]($backend).
+ * @internal
+ */
+function mapId64<R>(
+  idContainer:
+  | Id64String
```

Comment:
I have never seen a type declaration that begins with a pipe character. Is this short-hand for `undefined | ...`?

Reply:
`| A | B` is equivalent to `A|B`, you can always prepend a leading `|` to a type union in typescript. It's used to make multi-line unions prettier generally. Originally I supported more types in this function and I never collapsed the lines

Reply:
I'd collapse them and remove confusing extra pipe operator.

Reply:
I'm surprised you've never seen it before but if you haven't I'll take that as evidence it's much less common than i thought and remove it

Reply:
I never explicitly type this myself. I believe the argument list was unwieldy so I used vscode's format selection to run the popular prettier javascript formatter on the argument list and it spread the long union across several lines with the leading pipe. I've removed it now.